### PR TITLE
docker/docker#27484-check if sysctls are used in host network mode.

### PR DESCRIPTION
**Pull request by gaocegege**
ref docker/docker#27484

> There are two checks we can make for seeing if the user specified the host namespace.
>
> 1. The Namespaces list does not have the NEWNET struct added to it
> 2. If it does have a NEWNET struct added and the path is not an empty string then we need to
>   1. Do a readlink on the current processes network namespace os.Readlink("/proc/self/ns/net") 
>   2. Do a readlink on the Path provided in the struct and compare the current processes network namespace with the one provided in the Path. If they match then it should be considered as the host namespace.

Signed-off-by: Ce Gao <ce.gao@outlook.com>

### CODE REVIEW

**Comment by gaocegege** [comment]
PTAL @crosbymichael

**Comment by gaocegege** [comment]
Sorry about it, I have added it.

**Comment by crosbymichael** [comment]
LGTM

File: libcontainer/configs/validate/validator.go [hunk 1]
```diff
@@ -125,11 +127,26 @@ func (v *ConfigValidator) sysctl(config *configs.Config) error {
...
+				destOfCurrentProcess, err := os.Readlink(currentProcessNetns)
+				if err != nil {
+					return fmt.Errorf("read soft link %q error", currentProcessNetns)
+				}
+				// readlink on the Path provided in the struct
+				destOfContainer, err := os.Readlink(path)
+				if err != nil {
+					return fmt.Errorf("read soft link %q error", path)
+				}
+				if destOfContainer == destOfCurrentProcess {
```
Comment (mrunalp) [inline libcontainer/configs/validate/validator.go]: Can we add a helper method to compare namespace links?

Comment (gaocegege) [inline libcontainer/configs/validate/validator.go]: OK, the logic is a little complex.

**Comment by gaocegege** [comment]
PTAL

File: libcontainer/configs/validate/validator.go [hunk 2]
```diff
@@ -125,14 +125,37 @@ func (v *ConfigValidator) sysctl(config *configs.Config) error {
 			}
 		}
 		if strings.HasPrefix(s, "net.") {
-			if config.Namespaces.Contains(configs.NEWNET) {
-				continue
-			} else {
+			if !config.Namespaces.Contains(configs.NEWNET) {
 				return fmt.Errorf("sysctl %q is not allowed in the hosts network namespace", s)
 			}
+			if config.Namespaces.PathOf(configs.NEWNET) != "" {
```
Comment (cyphar) [inline libcontainer/configs/validate/validator.go]: You can do this:

```
if path := config.Namespaces.PathOf(configs.NEWNET); path != "" {
  if err := checkHostNs(s, config.Namespaces.PathOf(configs.NEWNET)); err != nil {
    return err
  }
}
```

Comment (gaocegege) [inline libcontainer/configs/validate/validator.go]: got it.

**Comment by cyphar** [comment]
LGTM

**Comment by gaocegege** [comment]
PTAL

**Comment by hqhq** [comment]
LGTM

**Comment by justincormack** [comment]
@gaocegege  I was testing an update to `runc` in docker and it looks like this causes a CI failure with:
```
14:08:40 
14:08:40 ----------------------------------------------------------------------
14:08:40 FAIL: docker_cli_run_unix_test.go:888: DockerSuite.TestRunSysctls
14:08:40 
14:08:40 docker_cli_run_unix_test.go:893:
14:08:40     out, _ := dockerCmd(c, "run", "--sysctl", "net.ipv4.ip_forward=1", "--name", "test", "busybox", "cat", "/proc/sys/net/ipv4/ip_forward")
14:08:40 docker_utils.go:470:
14:08:40     c.Assert(result, icmd.Matches, icmd.Success)
14:08:40 ... result *cmd.Result = &cmd.Result{Cmd:(*exec.Cmd)(0xc42044e840), ExitCode:125, Error:(*exec.ExitError)(0xc420bafca0), Timeout:false, outBuffer:(*cmd.lockedBuffer)(0xc4209ced80), errBuffer:(*cmd.lockedBuffer)(0xc4209cee10)} ("\nCommand: /go/src/github.com/docker/docker/bundles/1.13.0-dev/binary-client/docker run --sysctl net.ipv4.ip_forward=1 --name test busybox cat /proc/sys/net/ipv4/ip_forward\nExitCode: 125, Error: exit status 125\nStdout: \nStderr: sysctl \"net.ipv4.ip_forward\" is not in a separate kernel namespace\n/go/src/github.com/docker/docker/bundles/1.13.0-dev/binary-client/docker: Error response from daemon: oci runtime error: sysctl \"net.ipv4.ip_forward\" is not in a separate kernel namespace.\ntime=\"2016-10-25T14:08:40Z\" level=error msg=\"error getting events from daemon: net/http: request canceled\" \n\n")
14:08:40 ... expected cmd.Expected = cmd.Expected{ExitCode:0, Timeout:false, Error:"", Out:"", Err:""}
14:08:40 ... 
14:08:40 Command: /go/src/github.com/docker/docker/bundles/1.13.0-dev/binary-client/docker run --sysctl net.ipv4.ip_forward=1 --name test busybox cat /proc/sys/net/ipv4/ip_forward
14:08:40 ExitCode: 125, Error: exit status 125
14:08:40 Stdout: 
14:08:40 Stderr: sysctl "net.ipv4.ip_forward" is not in a separate kernel namespace
14:08:40 /go/src/github.com/docker/docker/bundles/1.13.0-dev/binary-client/docker: Error response from daemon: oci runtime error: sysctl "net.ipv4.ip_forward" is not in a separate kernel namespace.
14:08:40 time="2016-10-25T14:08:40Z" level=error msg="error getting events from daemon: net/http: request canceled" 
14:08:40 
14:08:40 
14:08:40 Failures:
14:08:40 ExitCode was 125 expected 0
14:08:40 Expected no error
14:08:40 
14:08:40 
14:08:40 
```

File: libcontainer/configs/validate/validator.go [hunk 3]
```diff
@@ -125,14 +125,36 @@ func (v *ConfigValidator) sysctl(config *configs.Config) error {
 			}
 		}
 		if strings.HasPrefix(s, "net.") {
-			if config.Namespaces.Contains(configs.NEWNET) {
-				continue
```
Comment (cyphar) [inline libcontainer/configs/validate/validator.go]: Ah yeah, this is a bug. I'll make a PR to fix it.

**Comment by gaocegege** [comment]
Ah, sorry about it. It seems that I lost a `continue`.

😟

**Pull request merged**